### DISCUSSION TITLE
How to report/flush open spans at end of application

### DISCUSSION THREAD

**Discussion by wcarmon** [tokio-rs/tracing#2463] [Q&A]
At the end of my `main` fn, I noticed the root span wasn't being reported (ended).  (probably because they report on a background thread)

I added a brief delay to allow some time, but it's hacky

```rust
fn main() {
    do_the_important_stuff();

    // -- give app some time to report any open spans
    sleep(Duration::from_millis(250));
}
```

For OpenTelemetry, I see [`opentelemetry::global::shutdown_tracer_provider();`](https://docs.rs/opentelemetry/latest/opentelemetry/global/fn.shutdown_tracer_provider.html), that is otel specific though.

Is there a `tracing` way to flush/cleanup?

**Comment by davidbarsky** [comment]
The core `tracing` libraries aren't really aware of writers beyond [`MakeWriter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html) inside of `tracing-subscriber`. There's a lot of ways to tackle this problem that's idiosyncratic to each Layer/Subscriber, runtime, and constraints (e.g., `std`/`no_std`) and I don't think it's the best use of time/resources for `tracing` to focus on this _now_.

That being said, depending on _what_ writer you're using (e.g., `tracing-appender`), you can use https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/struct.WorkerGuard.html to ensure flushes, but if you're writing to `stdout`/`stderr`, you really shouldn't be seeing dropped spans.
